### ISSUE TITLE
ObjectFormatter leads to stack overflow error

### ISSUE THREAD

**Issue by perqin** [Muyangmin/Android-PLog#30]
## Basic info
PLog version:  2.0.0-beta5

## Figure out a bug

How to reproduce:  

1. Create Java classes with cycle references or simply create a singleton.
2. Log the object or singleton.

Example code here:

```
public class A {
  public B b;
}

public class B {
  public A a;
}

public class Singleton {
  private static Singleton instance;

  public static Singleton getInstance() {
    if (instance == null) {
      instance = new Singleton();
    }
  }

  private Singleton() {}
}

public class Main {
  public void someMethod() {
    A a = new A();
    B b = new B();
    a.b = b;
    b.a = a;
    // Leads to StackOverflowError
    PLog.d("Object %s", a);
    // Leads to StackOverflowError too!
    PLog.d("Singleton %s", Singleton.getInstance());
  }
}
```

What expected:  

Actually I can only expect PLog to log the object without any exception. I've dug into the source code and find that the DefaultFormatter uses ObjectFormatter, which will log all fields of a object. In the example above, the ObjectFormatter will try to format `A.b`, which is also an object. Then it try to format `B.a`, and then `A.b`, and finally, boom! In the singleton example, the static field will be formatted too, so it also cause infinite loop.

It is not easy to log an object in pretty manner. Personally, I would suggest adding a `depth` option to the ObjectFormatter to limit the length of the formatted object references chain. Another choice would be extracting the primitive fields only.

Actually result:  

Stack overflow.

**Comment by Muyangmin** [comment]
Thanks for your feedback! 

Extract only primitive fields maybe not a good choice since it limit the library too much; a `depth` option or *cycling reference detecting* should work.

**Comment by Muyangmin** [comment]
This is already resolved by commit 13152a3, as referenced above. Now this feature is available in public stable release [v2.0.0](https://github.com/JumeiRdGroup/Android-PLog/releases/tag/v2.0.0), and this issue will be closed.

Thanks @perqin again for your feedback and suggestion!
